core/include/kernel/notif.h

```diff
@@ -11,19 +11,100 @@
 #include <tee_api_types.h>
 #include <types_ext.h>
 
+/*
+ * Notification values are divided into two kinds, asynchronous and
+ * synchronous, where the asynchronous has the lowest values.
+ * They are ordered as:
+ * 0			    Do bottom half
+ * 1..NOTIF_ASYNC_MAX	    Free for signalling in PTAs and should be
+ *			    allocated with notif_alloc_async_value()
+ * NOTIF_SYNC_VALUE_BASE..  Used as NOTIF_SYNC_VALUE_BASE + thread_id
+ * NOTIF_VALUE_MAX	    for mutex and condvar wait/wakeup
+ *
+ * Any value can be signalled with notif_send_sync() while only the ones
+ * <= CFG_CORE_ASYNC_NOTIF_MAX can be signalled with notif_send_async().
+ */
+
+#if CFG_CORE_ASYNC_NOTIF_INTID
+#define NOTIF_ASYNC_VALUE_MAX		U(63)
+#define NOTIF_SYNC_VALUE_BASE		(NOTIF_ASYNC_VALUE_MAX + U(1))
+#else
 #define NOTIF_SYNC_VALUE_BASE		0
+#endif
 
 #define NOTIF_VALUE_MAX			(NOTIF_SYNC_VALUE_BASE + \
 					 CFG_NUM_THREADS)
 
+#define NOTIF_VALUE_DO_BOTTOM_HALF	0
+
+enum notif_event {
+	NOTIF_EVENT_STARTED,
+	NOTIF_EVENT_DO_BOTTOM_HALF,
+	NOTIF_EVENT_STOPPED,
+};
+
+struct notif_driver {
+	void (*atomic_cb)(struct notif_driver *ndrv, enum notif_event ev);
+	void (*yielding_cb)(struct notif_driver *ndrv, enum notif_event ev);
```

Comment:
Some comments would be welcome. For example on `atomic_cb`: `/* Called with exception and interrupts masked and notification spinlock held */`. Maybe also mention which values of `ev` are handled by which callback?

Reply:
I'll fix.